mysys/my_largepage.c

```diff
@@ -79,88 +66,354 @@ void my_large_free(uchar* ptr)
+  size_t large_page_size= 0;
   DBUG_ENTER("my_large_malloc_int");
 
-  /* Align block size to my_large_page_size */
-  size= MY_ALIGN(size, (size_t) my_large_page_size);
-  
-  shmid = shmget(IPC_PRIVATE, size, SHM_HUGETLB | SHM_R | SHM_W);
-  if (shmid < 0)
+  while (1)
   {
-    if (my_flags & MY_WME)
-      fprintf(stderr,
-              "Warning: Failed to allocate %lu bytes from HugeTLB memory."
-              " errno %d\n", (ulong) size, errno);
+    mapflag= MAP_PRIVATE | MAP_ANONYMOUS;
+    if (my_use_large_pages)
+    {
+      large_page_size= my_next_large_page_size(*size, &page_i);
+      if (large_page_size)
+      {
+#ifdef HAVE_LINUX_LARGE_PAGES
+        mapflag|= MAP_HUGETLB | my_bit_size_t_log2(large_page_size) << MAP_HUGE_SHIFT;
+#else
+        mapflag|= MAP_ALIGNED_SUPER | MAP_ALIGNED(my_bit_size_t_log2(large_page_size));
+#endif
+      }
+    }
+    /* mmap adjusts the size to the hugetlb page size so no adjustment is needed */
+    ptr= mmap(NULL, *size, PROT_READ | PROT_WRITE, mapflag, -1, 0);
+    if (ptr == (void*) -1)
+    {
+      ptr= NULL;
+      if (my_flags & MY_WME)
+      {
+        if (large_page_size)
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from HugeTLB memory"
+                  "(page size %zu). errno %d\n", *size, large_page_size, errno);
+        }
+        else
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from memory."
+                  " errno %d\n", *size, errno);
+        }
+      }
+      /* try next smaller memory size */
+      if (large_page_size && errno == ENOMEM)
+        continue;
 
-    DBUG_RETURN(NULL);
+      /* other errors are more serious */
+      DBUG_RETURN(NULL);
+    }
+    else /* success */
+    {
+      if (large_page_size)
+      {
+         /* we do need to record the adjustment so that munmap gets called with
+            the right size. This is only the case for HUGETLB pages.
+            Align block size to large_page_size */
+         *size = MY_ALIGN(*size, (size_t) large_page_size);
+      }
+      DBUG_RETURN(ptr);
+    }
+    if (large_page_size == 0)
+    {
+      break; /* no more options to try */
+    }
   }
+  DBUG_RETURN(ptr);
+}
+
+#endif /* defined(HAVE_LINUX_LARGE_PAGES) || defined(HAVE_MMAP_ALIGNED) */
+
+#if defined(HAVE_GETPAGESIZES) && !defined(HAVE_LINUX_LARGE_PAGES)
+void my_get_large_page_sizes(size_t sizes[my_large_page_sizes_length])
+{
+  int nelem;
+
+  nelem = getpagesizes(NULL, 0);
+
+  assert(nelem <= my_large_page_sizes_length);
+  getpagesizes(sizes, my_large_page_sizes_length);
+  qsort(sizes, nelem, sizeof(size_t), size_t_cmp);
+  if (nelem < my_large_page_sizes_length)
+  {
+    sizes[nelem]= 0;
+  }
+}
+#endif
+
+#if defined(HAVE_MMAP) && !defined(_WIN32)
 
-  ptr = (uchar*) shmat(shmid, NULL, 0);
-  if (ptr == (uchar *) -1)
+/* mmap and Linux-specific large pages deallocator */
+
+my_bool my_large_free_int(void *ptr, size_t size)
```

Comment:
For manual poisoned memory ASAN requires that it will be unpoisoned manually too. `munmap()` certainly doesn't unpoison any memory. https://github.com/llvm/llvm-project/blob/master/compiler-rt/include/sanitizer/asan_interface.h#L21

IMO, ideally, the guy who poisoned memory should unpoison it too. But he might forget to do it. In that case unpoisoning memory before `munmap()` will help preventing false positives from ASAN.

@grooverdan what do you think, is it worth to unpoison memory here?

Reply:
Here is a relevant MariaDB commit https://github.com/MariaDB/server/commit/f6b58d29163adcb3b3fe1b48c56c9b80666f0b88

Reply:
Getting definitions in order:
<pre>UNIV_MEM_FREE -> MEM_NOACCESS -> ASAN_POISON_MEMORY_REGION
UNIV_MEM_ALLOC -> UNIV_MEM_INVALID -> MEM_UNDEFINED -> ASAN_UNPOISON_MEMORY_REGION
</pre>
That MEM_FREE definition really played with my head.

So yes, the current implementation seems to unpoision on allocation and freeing and this should too. I'm going to add that to the current patch sequence in this PR. Thanks for the references.

Reply:
IMO, those `MEM_` AND `UNIV_` macros exists for code obfuscation only..